core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala

```diff
@@ -0,0 +1,115 @@
+import software.amazon.awssdk.regions.{Region => AwsRegion}
+import software.amazon.awssdk.services.s3.model.{
+  HeadBucketRequest,
+  NoSuchBucketException,
+  S3Exception
+}
+import cats.Eq
+import cats.data.EitherT
+import cats.effect.{Async, ConcurrentEffect, ContextShift, Resource, Sync, Timer}
+import cats.instances.int._
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.functor._
+import eu.timepit.refined.auto._
+import monix.catnap.syntax._
+import scalaz.NonEmptyList
+
+object S3DestinationModule extends DestinationModule {
+  // Minimum 10MiB multipart uploads
+  private val PartSize = 10 * 1024 * 1024
+  private val Redacted = "<REDACTED>"
+  private val RedactedCreds = S3Credentials(AccessKey(Redacted), SecretKey(Redacted), Region(Redacted))
+
+  def destinationType = DestinationType("s3", 1L)
+
+  def sanitizeDestinationConfig(config: Json) = config.as[S3Config].result match {
+    case Left(_) => Json.jEmptyObject // don't expose credentials, even if we fail to decode the configuration.
+    case Right(cfg) => cfg.copy(credentials = RedactedCreds).asJson
+  }
+
+  def destination[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](
+      config: Json)
+      : Resource[F, Either[InitializationError[Json], Destination[F]]] = {
+
+    val configOrError = config.as[S3Config].toEither.leftMap {
+      case (err, _) =>
+        DestinationError.malformedConfiguration((destinationType, config, err))
+    }
+
+    val sanitizedConfig = sanitizeDestinationConfig(config)
+
+    (for {
+      cfg <- EitherT(Resource.pure[F, Either[InitializationError[Json], S3Config]](configOrError))
+      client <- EitherT(mkClient(cfg).map(_.asRight[InitializationError[Json]]))
+      upload = DefaultUpload(client, PartSize)
+      _ <- EitherT(Resource.liftF(isLive(client, sanitizedConfig, cfg.bucket)))
+    } yield (S3Destination(cfg.bucket, upload): Destination[F])).value
+  }
+
+  private def isLive[F[_]: Async](client: S3AsyncClient, originalConfig: Json, bucket: Bucket)
+      : F[Either[InitializationError[Json], Unit]] =
+    Async[F].delay(client.headBucket(HeadBucketRequest.builder.bucket(bucket.value).build))
+      .futureLift.as(().asRight[InitializationError[Json]]) recover {
+        case (_: NoSuchBucketException) =>
+          DestinationError.invalidConfiguration((
+            destinationType,
+            originalConfig,
+            NonEmptyList("Bucket does not exist"))).asLeft
+        // eq syntax is buggy in this case. Don't use it. It will lock-up the thread handling the request
+        // in slamdata-backend
+        case (e: S3Exception) if Eq[Int].eqv(403, e.statusCode) =>
+          DestinationError.accessDenied((
+            destinationType,
+            originalConfig,
+            "Access denied")).asLeft
+      }
+
+  private def mkClient[F[_]: Sync](cfg: S3Config): Resource[F, S3AsyncClient] = {
+    val client =
+      Sync[F].delay(
```

Comment:
I tried to configure a custom `ExecutorService` here, but it doesn't seem to work. I'm still seeing the default threads all around in addition to threads from my executor. [Thread dump](https://gist.github.com/jsantos17/e4049c163f6e2a5d2da70ee274dfa8b1). Note now we have both `s3-async-dest-1` and `aws-java-sdk-NettyEventLoop-0-7` 😕

You can find the implementation with a custom executor service [here](https://github.com/jsantos17/quasar-destination-s3/blob/ch8288-s3-destination-custom-executor/core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala#L108).